File: app/typescript/src/domain/production-plan/requirement.repository.ts
import { PrismaClient } from '@prisma/client'

// 所要情報の入力型
export interface CreateRequirementInput {
  requirementNumber: string
  orderId: number
  itemCode: string
  dueDate: Date
  requiredQuantity: number
  locationCode: string
}

// 所要情報の出力型
export interface Requirement {
  id: number
  requirementNumber: string
  orderId: number
  itemCode: string
  dueDate: Date
  requiredQuantity: number
  allocatedQuantity: number
  shortageQuantity: number
  locationCode: string
  createdAt: Date
  updatedAt: Date
}

export class RequirementRepository {
  constructor(private prisma: PrismaClient) {}

  async create(input: CreateRequirementInput): Promise<Requirement> {
    const result = await this.prisma.$queryRaw<Requirement[]>`
      INSERT INTO requirements (
        requirement_number, order_id, item_code, due_date,
        required_quantity, location_code
      )
      VALUES (
        ${input.requirementNumber}, ${input.orderId}, ${input.itemCode},
        ${input.dueDate}, ${input.requiredQuantity}, ${input.locationCode}
      )
      RETURNING
        id,
        requirement_number as "requirementNumber",
        order_id as "orderId",
        item_code as "itemCode",
        due_date as "dueDate",
        required_quantity as "requiredQuantity",
        allocated_quantity as "allocatedQuantity",
        shortage_quantity as "shortageQuantity",
        location_code as "locationCode",
        created_at as "createdAt",
        updated_at as "updatedAt"
    `
    return result[0]
  }

  async findByNumber(requirementNumber: string): Promise<Requirement | null> {
    const result = await this.prisma.$queryRaw<Requirement[]>`
      SELECT
        id,
        requirement_number as "requirementNumber",
        order_id as "orderId",
        item_code as "itemCode",
        due_date as "dueDate",
        required_quantity as "requiredQuantity",
        allocated_quantity as "allocatedQuantity",
        shortage_quantity as "shortageQuantity",
        location_code as "locationCode",
        created_at as "createdAt",
        updated_at as "updatedAt"
      FROM requirements
      WHERE requirement_number = ${requirementNumber}
    `
    return result.length > 0 ? result[0] : null
  }

  async findByOrderId(orderId: number): Promise<Requirement[]> {
    const result = await this.prisma.$queryRaw<Requirement[]>`
      SELECT
        id,
        requirement_number as "requirementNumber",
        order_id as "orderId",
        item_code as "itemCode",
        due_date as "dueDate",
        required_quantity as "requiredQuantity",
        allocated_quantity as "allocatedQuantity",
        shortage_quantity as "shortageQuantity",
        location_code as "locationCode",
        created_at as "createdAt",
        updated_at as "updatedAt"
      FROM requirements
      WHERE order_id = ${orderId}
      ORDER BY due_date
    `
    return result
  }

  async updateAllocated(
    id: number,
    allocatedQuantity: number,
    shortageQuantity: number
  ): Promise<Requirement> {
    const result = await this.prisma.$queryRaw<Requirement[]>`
      UPDATE requirements
      SET
        allocated_quantity = ${allocatedQuantity},
        shortage_quantity = ${shortageQuantity},
        updated_at = CURRENT_TIMESTAMP
      WHERE id = ${id}
      RETURNING
        id,
        requirement_number as "requirementNumber",
        order_id as "orderId",
        item_code as "itemCode",
        due_date as "dueDate",
        required_quantity as "requiredQuantity",
        allocated_quantity as "allocatedQuantity",
        shortage_quantity as "shortageQuantity",
        location_code as "locationCode",
        created_at as "createdAt",
        updated_at as "updatedAt"
    `
    return result[0]
  }
}

File: app/csharp/src/ProductionManagement.Infrastructure/Persistence/Repositories/RequirementRepository.cs
using Dapper;
using Npgsql;
using ProductionManagement.Application.Port.Out;
using ProductionManagement.Domain.Models.Plan;
using ProductionManagement.Infrastructure.Persistence.TypeHandlers;

namespace ProductionManagement.Infrastructure.Persistence.Repositories;

/// <summary>
/// 所要情報リポジトリ実装
/// </summary>
public class RequirementRepository : IRequirementRepository
{
    private readonly string _connectionString;

    static RequirementRepository()
    {
        SqlMapper.AddTypeHandler(new DateOnlyTypeHandler());
    }

    public RequirementRepository(string connectionString)
    {
        _connectionString = connectionString;
    }

    public async Task SaveAsync(Requirement requirement)
    {
        const string sql = """
            INSERT INTO "所要情報" (
                "所要NO", "オーダID", "品目コード", "納期",
                "必要数量", "引当済数量", "不足数量", "場所コード"
            ) VALUES (
                @RequirementNumber, @OrderId, @ItemCode, @DueDate,
                @RequiredQuantity, @AllocatedQuantity, @ShortageQuantity, @LocationCode
            )
            RETURNING "ID"
            """;

        await using var connection = new NpgsqlConnection(_connectionString);
        requirement.Id = await connection.ExecuteScalarAsync<int>(sql, new
        {
            requirement.RequirementNumber,
            requirement.OrderId,
            requirement.ItemCode,
            requirement.DueDate,
            requirement.RequiredQuantity,
            requirement.AllocatedQuantity,
            requirement.ShortageQuantity,
            requirement.LocationCode
        });
    }

    public async Task<Requirement?> FindByIdAsync(int id)
    {
        const string sql = """
            SELECT
                "ID" as Id,
                "所要NO" as RequirementNumber,
                "オーダID" as OrderId,
                "品目コード" as ItemCode,
                "納期" as DueDate,
                "必要数量" as RequiredQuantity,
                "引当済数量" as AllocatedQuantity,
                "不足数量" as ShortageQuantity,
                "場所コード" as LocationCode,
                "作成日時" as CreatedAt,
                "更新日時" as UpdatedAt
            FROM "所要情報"
            WHERE "ID" = @Id
            """;

        await using var connection = new NpgsqlConnection(_connectionString);
        return await connection.QuerySingleOrDefaultAsync<Requirement>(sql, new { Id = id });
    }

    public async Task<IReadOnlyList<Requirement>> FindByOrderIdAsync(int orderId)
    {
        const string sql = """
            SELECT
                "ID" as Id,
                "所要NO" as RequirementNumber,
                "オーダID" as OrderId,
                "品目コード" as ItemCode,
                "納期" as DueDate,
                "必要数量" as RequiredQuantity,
                "引当済数量" as AllocatedQuantity,
                "不足数量" as ShortageQuantity,
                "場所コード" as LocationCode,
                "作成日時" as CreatedAt,
                "更新日時" as UpdatedAt
            FROM "所要情報"
            WHERE "オーダID" = @OrderId
            ORDER BY "納期"
            """;

        await using var connection = new NpgsqlConnection(_connectionString);
        var result = await connection.QueryAsync<Requirement>(sql, new { OrderId = orderId });
        return result.ToList();
    }

    public async Task UpdateAllocationAsync(int id, decimal allocatedQuantity, decimal shortageQuantity)
    {
        const string sql = """
            UPDATE "所要情報"
            SET "引当済数量" = @AllocatedQuantity,
                "不足数量" = @ShortageQuantity,
                "更新日時" = CURRENT_TIMESTAMP
            WHERE "ID" = @Id
            """;

        await using var connection = new NpgsqlConnection(_connectionString);
        await connection.ExecuteAsync(sql, new { Id = id, AllocatedQuantity = allocatedQuantity, ShortageQuantity = shortageQuantity });
    }

    public async Task DeleteAllAsync()
    {
        const string sql = """DELETE FROM "所要情報" """;

        await using var connection = new NpgsqlConnection(_connectionString);
        await connection.ExecuteAsync(sql);
    }
}
